lib/graphql/schema/implementation/build_resolve.rb

```diff
@@ -0,0 +1,62 @@
+# frozen_string_literal: true
+# test_via: spec/graphql/schema/implementation_spec.rb
+module GraphQL
+  class Schema
+    class Implementation
+      # TODO, can users extend this, provide custom args?
+      SPECIAL_ARGS = Set.new([
+        :irep_node,
+        :ast_node,
+      ])
+
+      module BuildResolve
+        def self.build(impl_class, field)
+          field_name = field.name
+          field_args = field.arguments
+          # Remove camelization
+          method_name = field_name
+            .gsub(/([A-Z]+)([A-Z][a-z])/,'\1_\2')
+            .gsub(/([a-z\d])([A-Z])/,'\1_\2')
+            .downcase
+
+          if impl_class.method_defined?(method_name)
+            field_method = impl_class.instance_method(method_name)
+            graphql_args = []
+            special_args = []
+            field_method.parameters.each do |(type, name)|
+              case type
+              when :req, :opt
+                raise InvalidImplementationError.new(
+                  "positional arguments are not supported (use a keyword argument instead), see #{impl_class.name}##{method_name} (#{field_method.source_location.join(":")})"
+                )
+              when :key
+                raise InvalidImplementationError.new(
+                  "unexpected default value for #{name.inspect} (remove the Ruby default, one will be provided by GraphQL), see #{impl_class.name}##{method_name} (#{field_method.source_location.join(":")})"
+                )
+              when :keyreq
+                arg_name = name.to_s
+                if field_args.key?(arg_name)
+                  graphql_args << name
+                elsif SPECIAL_ARGS.include?(name)
+                  special_args << name
+                else
+                  raise InvalidImplementationError.new(
+                    "unexpected keyword #{name.inspect}, see #{impl_class.name}##{method_name} (#{field_method.source_location.join(":")})"
+                  )
+                end
+              end
+            end
```

Comment:
So I'm thinking about `:rest` and `:keyrest` args:
```
# :rest
def my_resolver(*args)
  # => args == [{ :some_argument1 => 'some value 1' }]
end

# :keyrest
def my_other_resolver(**kwargs)
  # => kwargs == { :some_argument1 => 'some value 1' }
end
```

A couple of thoughts on this:
- It seems like we would want to support those kinds of arguments. And if they exist, we'd basically only check to make sure that there are no positional args, and no unknown keyword args.
- Thinking through that case, if a user _did_ use `*args` or `**kwargs`, I'm not sure whether they would expect to get the special args or not.
- It seems like it'd be most predictable if we could guarantee that every call into a resolver method would be done with the exact same signature. But that would require us to take a stance that either all methods will accept `:ast_node` & friends, or no methods would accept them.

That third point is really about avoiding brittle scenarios, where changing from seemingly identical method signatures ends up changing behavior. For example:
```ruby
def my_resolver(some_arg1:, irep_node:, ast_node:)
end

# Several months later, someone refactors this into:
def my_resolver(**kwargs)
end
```

Perhaps the best option right now would be to raise an `InvalidImplementationError` upon encountering a `:rest` or `:keyrest` parameter.